hikari/impl/event_manager_base.py

```diff
@@ -50,46 +58,155 @@
+    def __event_types__(self) -> typing.Sequence[typing.Type[base_events.Event]]:
+        raise NotImplementedError
+
+
 def _default_predicate(_: event_manager.EventT_inv) -> bool:
     return True
 
 
+def filtered(
+    event_types: typing.Union[typing.Type[base_events.Event], typing.Sequence[typing.Type[base_events.Event]]],
+    cache_components: config.CacheComponents = config.CacheComponents.NONE,
+    /,
+) -> typing.Callable[[UnboundMethodT[EventManagerBaseT]], UnboundMethodT[EventManagerBaseT]]:
+    """Add metadata to a consumer method to indicate when it should be unmarshalled and dispatched.
+
+    Parameters
+    ----------
+    event_types
+        Types of the events this raw consumer method may dispatch.
+        This may either be a singular type of a sequence of types.
+
+    Other Parameters
+    ----------------
+    cache_components : hikari.config.CacheComponents
+        Bitfield of the cache components this event may make altering calls to.
+        This defaults to `hikari.config.CacheComponents.NONE`.
+    """
+    if isinstance(event_types, typing.Sequence):
+        # dict.fromkeys is used to remove any duplicate entries here
+        event_types = tuple(dict.fromkeys(itertools.chain.from_iterable(e.dispatches() for e in event_types)))
+
+    else:
+        event_types = event_types.dispatches()
+
+    def decorator(method: UnboundMethodT[EventManagerBaseT], /) -> UnboundMethodT[EventManagerBaseT]:
+        method.__cache_components__ = cache_components  # type: ignore[attr-defined]
+        method.__event_types__ = event_types  # type: ignore[attr-defined]
+        assert isinstance(method, _FilteredMethodT), "Incorrect attribute(s) set for a filtered method"
+        return method  # type: ignore[unreachable]
+
+    return decorator
+
+
+@attr.frozen()
+class _Consumer:
+    callback: ConsumerT
+    """The callback function for this consumer."""
+
+    cache_components: undefined.UndefinedOr[config.CacheComponents]
+    """Bitfield of the cache components this consumer makes modifying calls to, if set."""
+
+    event_types: undefined.UndefinedOr[typing.Sequence[typing.Type[base_events.Event]]]
+    """A sequence of the types of events this consumer dispatches to, if set."""
+
+
 class EventManagerBase(event_manager.EventManager):
     """Provides functionality to consume and dispatch events.
 
     Specific event handlers should be in functions named `on_xxx` where `xxx`
     is the raw event name being dispatched in lower-case.
     """
 
-    __slots__: typing.Sequence[str] = ("_app", "_listeners", "_consumers", "_waiters")
+    __slots__: typing.Sequence[str] = ("_app", "_consumers", "_enabled_consumers_cache", "_listeners", "_waiters")
 
     def __init__(self, app: traits.BotAware) -> None:
         self._app = app
-        self._consumers: typing.Dict[str, ConsumerT] = {}
+        self._consumers: typing.Dict[str, _Consumer] = {}
+        self._enabled_consumers_cache: typing.Dict[_Consumer, bool] = {}
```

Comment:
Could this cache be replaced with just `_Consumer` holding whether it is enabled or not? 

I just thought of this, but I dont know how well it will work.
Essentially, each consumer holds a value of how many listeners it has (0 for none) and whether the cache is enabled for it (this will be setup when the consumer is created)
Then, to check if we should dispatch or not, we check if the cache for it is enabled and if not, check the number of listeners it has.

Reply:
> Could this cache be replaced with just _Consumer holding whether it is enabled or not?

The thing with this idea is when registering listeners and waiters we don't easily know which consumer(s) which are relevant for the provided Event and to modify the cached value we'd have to somehow work out all the consumers an event class is relevant to which is possible but would add in a noticeable amount of complexity.

Chances are this'd mean we'd have to iterate over a resource each time listeners or waiters are added (likely just iterating over all the registered consumers and checking their `.event_types` to see if there's a match in them or another resource which stores something like `Dict[Set[Type[Event]], List[_Consumer]]`, really depends on which one ends up meaning less iteration and idk what the dynamics would be for that off the top of my head lol) but just thinking about this just gives me shivers.

Reply:
Using the set that I mentioned above, it should be quite easy to avoid all the loops, as all lookups to the event types will be O(1).

I don't like the temporary cache approach as it might be rendered useless in a lot of occasions (like a lot of waiters being constantly added and removed, causing the cache to be cleared all the time). I would like to try out the approach of making the consumers hold the information and just fetch it from there when needed, if that is fine with you.

Reply:
> Using the set that I mentioned above, it should be quite easy to avoid all the loops, as all lookups to the event types will be O(1).

And how would you deal with the fact that when registering or unregistering a listener or waiter you'd have to iterate over every event it's event dispatches for to change the state for any relevant consumer which is dispatching that event which would look something like

```py
for event_type in event.dispatches():
    for consumer in self._event_to_consumer[event_type]:
        consumer.listeners += 1
```

or

```py
for event_type in event.dispatches():
    for consumer in self._consumers:
        if event_type in consumer.event_types:
            consumer.listeners += 1
```

when adding a listener/waiter and -= when removing a listener or waiter

Reply:
> I don't like the temporary cache approach as it might be rendered useless in a lot of occasions (like a lot of waiters being constantly added and removed, causing the cache to be cleared all the time)

So in the case where a bot is registering 1 and only 1 waiter for an event type between every single event dispatch and that waiter presumably gets unregistered when the next event is received?